File: MiniDOMBase/DMD/Sistema/Calendar/Comparer/ActivePersonDateComparer.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace minidom
{
    public partial class Sistema
    {

        /// <summary>
        /// Comparatore di oggetti di tipo <see cref="CActivePerson"/>
        /// </summary>
        public class ActivePersonDateComparer 
            : IComparer, IComparer<CActivePerson>
        {

            /// <summary>
            /// Costruttore
            /// </summary>
            public ActivePersonDateComparer()
            {
                //DMDObject.IncreaseCounter(this);
            }

            private int CompareByCategoria(CActivePerson x, CActivePerson y)
            {
                if (string.IsNullOrEmpty(x.Categoria))
                    x.Categoria = "Normale";
                if (string.IsNullOrEmpty(y.Categoria))
                    y.Categoria = "Normale";
                var items = new string[] { "Urgente", "Importante", "Normale", "Poco importante" };
                int ix = DMD.Arrays.IndexOf(items, x.Categoria);
                int iy = DMD.Arrays.IndexOf(items, y.Categoria);
                return ix - iy;
            }

            private int CompareByNominativo(CActivePerson x, CActivePerson y)
            {
                return DMD.Strings.Compare(x.Nominativo, y.Nominativo, true);
            }

            private int CompareByData(CActivePerson x, CActivePerson y)
            {
                int ret = DMD.DateUtils.Compare(DMD.DateUtils.GetDatePart(x.Data), DMD.DateUtils.GetDatePart(y.Data));
                if (ret == 0)
                {
                    DateTime d1 = (DateTime)DMD.DateUtils.GetDatePart(x.Data);
                    DateTime d2 = (DateTime)DMD.DateUtils.GetDatePart(y.Data);
                    int diff1;
                    int diff2;
                    if (x.GiornataIntera)
                    {
                        if (y.GiornataIntera)
                        {
                            // x giornata intera e y giornata intera
                            ret = CompareByCategoria(x, y);
                            if (ret == 0)
                            {
                                if (d1 < d2)
                                {
                                    ret = -1;
                                }
                                else if (d1 > d2)
                                {
                                    ret = 1;
                                }
                                else
                                {
                                    ret = 0;
                                }
                            }
                        }
                        else
                        {
                            // x giornata intera e y non giornata intera
                            diff2 = (int)DMD.DateUtils.DateDiff(DMD.DateTimeInterval.Minute, DMD.DateUtils.Now(), y.Data.Value);
                            if (diff2 <= y.Promemoria)
                            {
                                ret = 1;
                            }
                            else if (d1 < d2)
                            {
                                ret = -1;
                            }
                            else
                            {
                                ret = 1;
                            }
                        }
                    }
                    else if (y.GiornataIntera)
                    {
                        // x non giornata intera e y giornata intera
                        diff1 = (int)DMD.DateUtils.DateDiff(DMD.DateTimeInterval.Minute, DMD.DateUtils.Now(), x.Data.Value);
                        if (diff1 <= x.Promemoria)
                        {
                            ret = -1;
                        }
                        else if (d1 < d2)
                        {
                            ret = -1;
                        }
                        else
                        {
                            ret = 1;
                        }
                    }
                    else if (x.Data < y.Data == true)
                    {
                        ret = -1;
                    }
                    else if (x.Data > y.Data == true)
                    {
                        ret = 1;
                    }
                    else
                    {
                        ret = 0;
                    }
                }

                return ret;
            }

            /// <summary>
            /// Compara due oggetti
            /// </summary>
            /// <param name="x"></param>
            /// <param name="y"></param>
            /// <returns></returns>
            public int Compare(CActivePerson x, CActivePerson y)
            {
                int ret = CompareByData(x, y);
                if (ret == 0)
                    ret = CompareByCategoria(x, y);
                if (ret == 0)
                    ret = CompareByNominativo(x, y);
                return ret;
            }

            int IComparer.Compare(object x, object y)
            {
                return Compare((CActivePerson)x, (CActivePerson)y);
            }

            //~ActivePersonDateComparer()
            //{
            //    //DMDObject.DecreaseCounter(this);
            //}
        }
    }
}

File: MiniDOMBase/DMD/Sistema/Calendar/ActivePersonDateComparer.vb
Imports minidom
Imports minidom.Databases
Imports minidom.Sistema
Imports minidom.Anagrafica

Partial Class Sistema


    Public Class ActivePersonDateComparer
        Implements IComparer

        Public Sub New()
            DMDObject.IncreaseCounter(Me)
        End Sub

        Private Function CompareByCategoria(ByVal x As CActivePerson, ByVal y As CActivePerson) As Integer
            If (x.Categoria = "") Then x.Categoria = "Normale"
            If (y.Categoria = "") Then y.Categoria = "Normale"
            Dim items() As String = {"Urgente", "Importante", "Normale", "Poco importante"}
            Dim ix As Integer = Arrays.IndexOf(items, x.Categoria)
            Dim iy As Integer = Arrays.IndexOf(items, y.Categoria)
            Return ix - iy
        End Function

        Private Function CompareByNominativo(ByVal x As CActivePerson, ByVal y As CActivePerson) As Integer
            Return Strings.Compare(x.Nominativo, y.Nominativo, CompareMethod.Text)
        End Function

        Private Function CompareByData(ByVal x As CActivePerson, ByVal y As CActivePerson) As Integer
            Dim ret As Integer = DateUtils.Compare(DateUtils.GetDatePart(x.Data), DateUtils.GetDatePart(y.Data))
            If (ret = 0) Then
                Dim d1 As Date = DateUtils.GetDatePart(x.Data)
                Dim d2 As Date = DateUtils.GetDatePart(y.Data)
                Dim diff1 As Integer
                Dim diff2 As Integer
                If (x.GiornataIntera) Then
                    If (y.GiornataIntera) Then
                        'x giornata intera e y giornata intera
                        ret = Me.CompareByCategoria(x, y)
                        If (ret = 0) Then
                            If (d1 < d2) Then
                                ret = -1
                            ElseIf (d1 > d2) Then
                                ret = 1
                            Else
                                ret = 0
                            End If
                        End If
                    Else
                        'x giornata intera e y non giornata intera
                        diff2 = DateUtils.DateDiff(DateInterval.Minute, DateUtils.Now, y.Data.Value)
                        If (diff2 <= y.Promemoria) Then
                            ret = 1
                        Else
                            If (d1 < d2) Then
                                ret = -1
                            Else
                                ret = 1
                            End If
                        End If
                    End If
                Else
                    If (y.GiornataIntera) Then
                        'x non giornata intera e y giornata intera
                        diff1 = DateUtils.DateDiff(DateInterval.Minute, DateUtils.Now, x.Data.Value)
                        If (diff1 <= x.Promemoria) Then
                            ret = -1
                        Else
                            If (d1 < d2) Then
                                ret = -1
                            Else
                                ret = 1
                            End If
                        End If
                    Else
                        If (x.Data < y.Data) Then
                            ret = -1
                        ElseIf (x.Data > y.Data) Then
                            ret = 1
                        Else
                            ret = 0
                        End If
                    End If
                End If
            End If

            Return ret
        End Function

        Public Function Compare(x As CActivePerson, y As CActivePerson) As Integer
            Dim ret As Integer = Me.CompareByData(x, y)
            If (ret = 0) Then ret = Me.CompareByCategoria(x, y)
            If (ret = 0) Then ret = Me.CompareByNominativo(x, y)
            Return ret
        End Function

        Private Function Compare1(x As Object, y As Object) As Integer Implements IComparer.Compare
            Return Me.Compare(x, y)
        End Function

        Protected Overrides Sub Finalize()
            MyBase.Finalize()
            DMDObject.DecreaseCounter(Me)
        End Sub
    End Class


End Class